File: DEMO/Demos/13_ChangingAtRuntime/Delphi/Unit1.pas
unit Unit1;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, vgr_ColorButton, pr_Common, pr_Classes, Db, DBTables;

type
  TForm1 = class(TForm)
    Button1: TButton;
    Button2: TButton;
    Label1: TLabel;
    Label2: TLabel;
    Label4: TLabel;
    bBackColor: TvgrColorButton;
    bTextColor: TvgrColorButton;
    Label5: TLabel;
    prReport1: TprReport;
    Table1: TTable;
    edName: TComboBox;
    procedure Button2Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure prReport1FirstPassObject(Sender: TObject; Obj: TprObj;
      var ManuallyProcessed: Boolean);
    procedure Button1Click(Sender: TObject);
  private
    { Private declarations }
    FOnFirstPassObjectUsed: Boolean;
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}

procedure TForm1.Button2Click(Sender: TObject);
var
  AMemoObj: TprMemoObj;
begin
  AMemoObj := TprMemoObj(prReport1.FindObject(edName.Text));
  if not (AMemoObj is TprMemoObj) then
  begin
    ShowMessage(Format('Object [%s] not found', [edName.Text]));
    exit;
  end;

  with AMemoObj.DefVersion do
  begin
    FillColor := bBackColor.SelectedColor;
    Font.Color := bTextColor.SelectedColor;
  end;

  FOnFirstPassObjectUsed := False;
  prReport1.PrepareReport;
  prReport1.PreviewPreparedReport(True);
end;

procedure TForm1.FormCreate(Sender: TObject);
var
  I: Integer;
begin
  for I := 0 to prReport1.ObjectCount - 1 do
    if prReport1.Objects[I] is TprMemoObj then
      edName.Items.Add(Format('%s', [prReport1.Objects[I].Name, TprMemoObj(prReport1.Objects[I]).DefVersion.Memo.Text]));
  if edName.Items.Count > 0 then
    edName.Text := edName.Items[0];
end;

procedure TForm1.prReport1FirstPassObject(Sender: TObject; Obj: TprObj;
  var ManuallyProcessed: Boolean);
begin
  if (Obj.Name = edName.Text) and (Obj is TprMemoObj) and FOnFirstPassObjectUsed then
    with TprMemoObj(Obj).GenCurVersion do
    begin
      FillColor := bBackColor.SelectedColor;
      Font.Color := bTextColor.SelectedColor;
    end;
end;

procedure TForm1.Button1Click(Sender: TObject);
begin
  FOnFirstPassObjectUsed := True;
  prReport1.PrepareReport;
  prReport1.PreviewPreparedReport(True);
end;

end.

File: DEMO/Demos/13_ChangingAtRuntime/Builder/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#include <pr_Classes.hpp>
#include <pr_Common.hpp>
#include <vgr_ColorButton.hpp>
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  for (int i = 0; i < prReport1->ObjectCount; i++)
    if (dynamic_cast<TprMemoObj*>(prReport1->Objects[i]) != NULL)
      edName->Items->Add(dynamic_cast<TprMemoObj*>(prReport1->Objects[i])->Name);
  if (edName->Items->Count > 0)
    edName->Text = edName->Items->Strings[0];
}
//---------------------------------------------------------------------------

void __fastcall TForm1::prReport1FirstPassObject(TObject *Sender,
      TprObj *Obj, bool &ManuallyProcessed)
{
  TprMemoObj* m;

  if ((Obj->Name == edName->Text) && (dynamic_cast<TprMemoObj*>(Obj) != NULL) && FOnFirstPassObjectUsed)
  {
    m = (TprMemoObj*)Obj;
    m->GenCurVersion->FillColor = bBackColor->SelectedColor;
    m->GenCurVersion->Font->Color = bTextColor->SelectedColor;
  }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button1Click(TObject *Sender)
{
  FOnFirstPassObjectUsed = true;
  prReport1->PrepareReport();
  prReport1->PreviewPreparedReport(true);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button2Click(TObject *Sender)
{
  TprMemoObj* m;

  m = dynamic_cast<TprMemoObj*>(prReport1->FindObject(edName->Text));
  if (m != NULL)
  {
    ShowMessage("Object [" + edName->Text + "] not found");
    return;
  }

  m->DefVersion->FillColor = bBackColor->SelectedColor;
  m->DefVersion->Font->Color = bTextColor->SelectedColor;

  FOnFirstPassObjectUsed = false;
  prReport1->PrepareReport();
  prReport1->PreviewPreparedReport(true);
}
//---------------------------------------------------------------------------
